File: zertz-core/src/game.rs
use std::cell::Cell;
use std::fmt::{self, Debug, Display};

use serde::{Deserialize, Serialize};

// macro include
use crate::append_occupied_coordinate;

use crate::{
    board::*,
    coordinate::{Coordinate, CoordinateIter, Direction},
    error::{self, ZertzCoreError},
    union_find::UnionFind,
};

const MAIN_EMPTY_COORD: Coordinate = Coordinate::new(8, 0);

#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize)]
pub(crate) struct MarbleCount {
    white_count: usize,
    gray_count: usize,
    black_count: usize,
}

impl MarbleCount {
    fn inc(&mut self, marble: Marble) {
        match marble {
            Marble::White => self.white_count += 1,
            Marble::Gray => self.gray_count += 1,
            Marble::Black => self.black_count += 1,
        }
    }

    fn dec(&mut self, marble: Marble) -> bool {
        match marble {
            Marble::White => {
                if self.white_count == 0 {
                    return false;
                }
                self.white_count -= 1;
            }
            Marble::Gray => {
                if self.gray_count == 0 {
                    return false;
                }
                self.gray_count -= 1;
            }
            Marble::Black => {
                if self.black_count == 0 {
                    return false;
                }
                self.black_count -= 1;
            }
        }

        true
    }

    #[inline]
    fn is_win(&self) -> bool {
        self.white_count >= 4
            || self.gray_count >= 5
            || self.black_count >= 6
            || (self.white_count >= 3 && self.gray_count >= 3 && self.black_count >= 3)
    }
}

#[repr(u8)]
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub enum Player {
    Alice,
    Bob,
    Tie,
}

impl Player {
    fn change_player(&mut self) {
        match self {
            Self::Alice => *self = Self::Bob,
            Self::Bob => *self = Self::Alice,
            _ => {}
        }
    }
}

impl From<Player> for usize {
    fn from(player: Player) -> usize {
        match player {
            Player::Alice => 0,
            Player::Bob => 1,
            Player::Tie => 2,
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct CatchableMove {
    start_coord: Coordinate,
    catched_coord: Coordinate,
    marble_land_coord: Coordinate,
}

impl Display for CatchableMove {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{{{} -> {}}}", self.start_coord, self.marble_land_coord)
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub enum GameState {
    CheckIsCatchable,
    PutMarble,
    CatchMarble,
    GameEnd(Player),
}

#[derive(Serialize, Deserialize)]
pub struct Game {
    pub(crate) board: Board,
    pub(crate) board_replace_history: Vec<Board>,
    #[serde(skip)]
    pub(crate) components: UnionFind<Coordinate>,
    pub(crate) current_player: Player,
    pub(crate) game_state: GameState,
    pub(crate) players_score: [MarbleCount; 2],
    pub(crate) repeat_count: Cell<usize>,
    pub(crate) total_marble: MarbleCount,
}

// ╭──────────────────────────────────────────────────────────╮
// │                      Basic Game Api                      │
// ╰──────────────────────────────────────────────────────────╯

impl Game {
    pub(crate) fn new(kind: BoardKind) -> Self {
        let mut output = Self {
            board: Board::new(kind),
            board_replace_history: Vec::with_capacity(20),
            components: UnionFind::from(CoordinateIter::new().collect::<Vec<_>>()),
            current_player: Player::Alice,
            game_state: GameState::PutMarble,
            players_score: [MarbleCount::default(); 2],
            repeat_count: Cell::new(0),
            total_marble: MarbleCount {
                white_count: 6,
                gray_count: 8,
                black_count: 10,
            },
        };
        output.calculate_components();

        output
    }
}

// ╭──────────────────────────────────────────────────────────╮
// │                   Catching Marble Api                    │
// ╰──────────────────────────────────────────────────────────╯

impl Game {
    pub(crate) fn catch_marble(
        &mut self,
        catch_data: CatchableMove,
    ) -> error::Result<Option<Vec<CatchableMove>>> {
        let CatchableMove {
            start_coord,
            catched_coord,
            marble_land_coord,
        } = catch_data;

        match self.board[catched_coord] {
            Ring::Occupied(Marble::White) => {
                self.players_score[usize::from(self.current_player)].white_count += 1
            }
            Ring::Occupied(Marble::Gray) => {
                self.players_score[usize::from(self.current_player)].gray_count += 1
            }
            Ring::Occupied(Marble::Black) => {
                self.players_score[usize::from(self.current_player)].black_count += 1
            }
            _ => return Err(ZertzCoreError::FailedToCatchMarble),
        }

        self.board[marble_land_coord] = self.board[start_coord];
        self.board[start_coord] = Ring::Vacant;
        self.board[catched_coord] = Ring::Vacant;

        self.board_replace_history.push(self.board);

        let list_catchable = self.list_catchable_once(marble_land_coord);
        Ok(if list_catchable.is_empty() {
            self.current_player.change_player();
            self.game_state = if let Some(winner) = self.who_is_win(&self.board) {
                GameState::GameEnd(winner)
            } else {
                GameState::PutMarble
            };

            None
        } else {
            self.game_state = if let Some(winner) = self.who_is_win(&self.board) {
                GameState::GameEnd(winner)
            } else {
                GameState::CatchMarble
            };
            Some(list_catchable)
        })
    }

    pub(crate) fn list_all_catchable(&self) -> Vec<CatchableMove> {
        let mut output = Vec::with_capacity(81);

        for coord in CoordinateIter::new() {
            if let Ring::Occupied(_) = self.board[coord] {
                output.push(self.list_catchable_once(coord));
            }
        }

        output.into_iter().flatten().collect()
    }

    fn list_catchable_once(&self, coord: Coordinate) -> Vec<CatchableMove> {
        let mut output = Vec::with_capacity(6);

        append_occupied_coordinate!(self: output, coord, Direction::UP | Direction::RIGHT);
        append_occupied_coordinate!(self: output, coord, Direction::UP);
        append_occupied_coordinate!(self: output, coord, Direction::LEFT);
        append_occupied_coordinate!(self: output, coord, Direction::LEFT | Direction::DOWN);
        append_occupied_coordinate!(self: output, coord, Direction::DOWN);
        append_occupied_coordinate!(self: output, coord, Direction::RIGHT);

        output
    }
}

// ╭──────────────────────────────────────────────────────────╮
// │                    Putting Marble Api                    │
// ╰──────────────────────────────────────────────────────────╯

impl Game {
    pub(crate) fn put_marble(
        &mut self,
        put_coord: Coordinate,
        remove_coord: Coordinate,
        marble: Marble,
    ) -> error::Result<()> {
        if let Ring::Vacant = self.board[put_coord] {
            if !self.total_marble.dec(marble)
                && !self.players_score[usize::from(self.current_player)].dec(marble)
            {
                return Err(ZertzCoreError::InvalidPuttingMarble);
            }
            self.board[put_coord] = Ring::Occupied(marble);
        } else {
            return Err(ZertzCoreError::InvalidPuttingMarble);
        }

        self.remove_ring(remove_coord)?;
        self.remove_isolated_island();
        self.current_player.change_player();
        self.game_state = GameState::CheckIsCatchable;

        Ok(())
    }

    fn remove_ring(&mut self, coord: Coordinate) -> error::Result<()> {
        let list_removable = self.collect_removable_rings();

        if !list_removable.is_empty() {
            if !self.valid_to_remove_ring(coord) {
                return Err(ZertzCoreError::InvalidRingToRemove);
            }

            self.board_replace_history.clear();
            self.repeat_count.set(0);

            let ring = &mut self.board[coord];
            match ring {
                Ring::Vacant => *ring = Ring::Empty,
                _ => return Err(ZertzCoreError::InvalidRingToRemove),
            }
            self.calculate_components();
        }

        Ok(())
    }

    fn collect_removable_rings(&self) -> Vec<Coordinate> {
        let mut output = Vec::with_capacity(81);

        for coord in CoordinateIter::new() {
            if self.valid_to_remove_ring(coord) {
                output.push(coord);
            }
        }

        output
    }

    fn valid_to_remove_ring(&self, coord: Coordinate) -> bool {
        let up_right = coord.adjacent(Direction::UP | Direction::RIGHT);
        let up = coord.adjacent(Direction::UP);
        let left = coord.adjacent(Direction::LEFT);
        let left_down = coord.adjacent(Direction::LEFT | Direction::DOWN);
        let down = coord.adjacent(Direction::DOWN);
        let right = coord.adjacent(Direction::RIGHT);

        (matches!(self.board.get_option(up_right), None | Some(Ring::Empty))
            && matches!(self.board.get_option(up), None | Some(Ring::Empty)))
            || (matches!(self.board.get_option(left), None | Some(Ring::Empty))
                && matches!(self.board.get_option(up), None | Some(Ring::Empty)))
            || (matches!(self.board.get_option(left), None | Some(Ring::Empty))
                && matches!(self.board.get_option(left_down), None | Some(Ring::Empty)))
            || (matches!(self.board.get_option(down), None | Some(Ring::Empty))
                && matches!(self.board.get_option(left_down), None | Some(Ring::Empty)))
            || (matches!(self.board.get_option(down), None | Some(Ring::Empty))
                && matches!(self.board.get_option(right), None | Some(Ring::Empty)))
            || (matches!(self.board.get_option(up_right), None | Some(Ring::Empty))
                && matches!(self.board.get_option(right), None | Some(Ring::Empty)))
    }

    pub(crate) fn calculate_components(&mut self) {
        self.components.clear();

        let mut right_coord;
        let mut up_coord;
        let mut up_right_coord;
        for coord in CoordinateIter::new() {
            if self.board[coord] == Ring::Empty {
                self.components.union(&coord, &MAIN_EMPTY_COORD);
            }

            right_coord = coord.raw_adjacent(Direction::RIGHT);
            up_coord = coord.raw_adjacent(Direction::UP);
            up_right_coord = coord.raw_adjacent(Direction::RIGHT | Direction::UP);

            match self.board.get(right_coord) {
                Some(Ring::Empty) => self.components.union(&right_coord, &MAIN_EMPTY_COORD),
                Some(ring) if *ring == self.board[coord] => {
                    self.components.union(&coord, &right_coord)
                }
                _ => {}
            }

            match self.board.get(up_coord) {
                Some(Ring::Empty) => self.components.union(&up_coord, &MAIN_EMPTY_COORD),
                Some(ring) if *ring == self.board[coord] => {
                    self.components.union(&coord, &up_coord)
                }
                _ => {}
            }

            match self.board.get(up_right_coord) {
                Some(Ring::Empty) => self.components.union(&up_right_coord, &MAIN_EMPTY_COORD),
                Some(ring) if *ring == self.board[coord] => {
                    self.components.union(&coord, &up_right_coord)
                }
                _ => {}
            }
        }
    }

    fn remove_isolated_island(&mut self) {
        let main_components = self.components.components();

        let mut component_members;
        'components: for main_coord in main_components {
            if let Ring::Empty = self.board[main_coord] {
                continue;
            }

            component_members = Vec::with_capacity(81);

            for coord in CoordinateIter::new() {
                if self.components.find(&main_coord) != self.components.find(&coord) {
                    continue;
                }

                if let Ring::Vacant = self.board[coord] {
                    continue 'components;
                }

                component_members.push(coord);
            }

            for coord in component_members {
                match self.board[coord] {
                    Ring::Occupied(marble) => {
                        self.players_score[usize::from(self.current_player)].inc(marble);
                    }
                    _ => unreachable!(),
                }
                self.board[coord] = Ring::Empty;
                self.components.union(&coord, &MAIN_EMPTY_COORD);
            }
        }
    }
}

// ╭──────────────────────────────────────────────────────────╮
// │                     Check who is win                     │
// ╰──────────────────────────────────────────────────────────╯

impl Game {
    fn who_is_win(&self, board: &Board) -> Option<Player> {
        if self.players_score[0].is_win() {
            return Some(Player::Alice);
        } else if self.players_score[1].is_win() {
            return Some(Player::Bob);
        } else if self.board_replace_history.contains(board) {
            let count = self.repeat_count.get();
            self.repeat_count.set(count + 1);

            if self.repeat_count.get() >= 3 {
                return Some(Player::Tie);
            }
        }

        None
    }
}

impl Debug for Game {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "{:=<50}", "")?;
        write!(f, "\n{:-^63}", "\x1b[1m\x1b[97m <Board> \x1b[0m")?;
        write!(f, "{:?}", self.board)?;
        writeln!(
            f,
            "\n{:-^63}\n",
            "\x1b[1m\x1b[97m <Player Informations> \x1b[0m"
        )?;
        writeln!(f, "{:#?}", self.players_score)?;
        writeln!(f, "\n{:-^63}\n", "\x1b[1m\x1b[97m <Marble Leftoff> \x1b[0m")?;
        writeln!(f, "{:#?}", self.total_marble)?;
        writeln!(f, "\x1b[1m[{:#?} is playing]\x1b[0m", self.current_player)?;

        Ok(())
    }
}

File: zertz-core/Game.zig
const std = @import("std");
const coordinate = @import("./coordinate.zig");
const meta = std.meta;

const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;
const Board = @import("./board.zig").Board;
const BoardKind = @import("./board.zig").BoardKind;
const Coordinate = coordinate.Coordinate;
const CoordinateIter = coordinate.CoordinateIter;
const Direction = coordinate.Direction;
const Marble = @import("./board.zig").Marble;
const Ring = @import("./board.zig").Ring;
const UnionFind = @import("./union_find.zig").UnionFind;

const main_empty_coord = Coordinate.new(8, 0);

const MarbleCount = struct {
    white_count: usize,
    gray_count: usize,
    black_count: usize,

    fn new(white: usize, gray: usize, black: usize) @This() {
        return @This(){ .white_count = white, .gray_count = gray, .black_count = black };
    }

    fn inc(self: *@This(), marble: Marble) void {
        switch (marble) {
            inline else => |marble_name| @field(self, @tagName(marble_name) ++ "_count") += 1,
        }
    }

    fn dec(self: *@This(), marble: Marble) bool {
        switch (marble) {
            inline else => |marble_name| {
                var marble_count = &@field(self, @tagName(marble_name) ++ "_count");
                if (marble_count.* == 0) return false;
                marble_count.* -= 1;
            },
        }

        return true;
    }

    fn isWin(self: @This()) bool {
        return self.white_count >= 4 //
        or self.gray_count >= 5 //
        or self.black_count >= 6 //
        or (self.white_count >= 3 and self.gray_count >= 3 and self.black_count >= 3);
    }
};

pub const Player = enum(usize) {
    alice,
    bob,
    tie,

    fn changePlayer(self: *@This()) void {
        switch (self.*) {
            .alice => self.* = .bob,
            .bob => self.* = .alice,
            .tie => {},
        }
    }
};

pub const CatchableMove = struct {
    start_coord: Coordinate,
    catched_coord: Coordinate,
    marble_land_coord: Coordinate,
};

pub const GameState = union(enum) {
    check_is_catchable,
    found_sequential_move,
    put_marble,
    catch_marble,
    game_end: Player,
};

pub const GameError = std.mem.Allocator.Error || error{
    FailedToCatchMarble,
    InvalidPuttingMarble,
    InvalidRingToRemove,
};

// Fields for Game
allocator: Allocator,
board: Board,
board_replace_history: ArrayList(Board),
components: UnionFind(Coordinate),
current_player: Player,
game_state: GameState,
players_score: [2]MarbleCount,
repeat_count: usize,
total_marble: MarbleCount,
sequential_move_list: ?ArrayList(CatchableMove),
// END Fields

pub const Game = @This();
const Self = @This();

// ╭──────────────────────────────────────────────────────────╮
// │                      Basic Game Api                      │
// ╰──────────────────────────────────────────────────────────╯

pub fn init(allocator: Allocator, kind: BoardKind) GameError!Self {
    var output = Self{
        .allocator = allocator,
        .board = Board.new(kind),
        .current_player = .alice,
        .game_state = .put_marble,
        .players_score = [_]MarbleCount{ MarbleCount.new(0, 0, 0), MarbleCount.new(0, 0, 0) },
        .repeat_count = 0,
        .total_marble = MarbleCount.new(6, 8, 10),
        .sequential_move_list = null,
    };

    output.board_replace_history = try ArrayList(Board).initCapacity(allocator, 20);
    errdefer output.board_replace_history.deinit();

    var elems: [81]Coordinate = undefined;

    var iter = CoordinateIter.new();
    var i: usize = 0;
    while (iter.next()) |coord| {
        elems[i] = coord;
        i += 1;
    }

    output.components = try UnionFind(Coordinate).init(allocator, &elems);
    errdefer output.components.deinit();

    output.calculateCompoments();

    return output;
}

pub fn deinit(self: *Self) void {
    self.board_replace_history.deinit();
    self.components.deinit();
    if (self.sequential_move_list) |inner| {
        inner.deinit();
    }
}

// ╭──────────────────────────────────────────────────────────╮
// │                   Catching Marble Api                    │
// ╰──────────────────────────────────────────────────────────╯

pub fn catchMarble(self: *Self, catch_data: CatchableMove) GameError!void {
    const start_coord = catch_data.start_coord;
    const catched_coord = catch_data.catched_coord;
    const marble_land_coord = catch_data.marble_land_coord;

    switch (self.board.getRaw(catched_coord)) {
        .occupied => |marble| switch (marble) {
            .white => self.players_score[@enumToInt(self.current_player)].white_count += 1,
            .gray => self.players_score[@enumToInt(self.current_player)].gray_count += 1,
            .black => self.players_score[@enumToInt(self.current_player)].black_count += 1,
        },
        else => return error.FailedToCatchMarble,
    }

    self.board.getRawMut(marble_land_coord).* = self.board.getRaw(start_coord);
    self.board.getRawMut(start_coord).* = .vacant;
    self.board.getRawMut(catched_coord).* = .vacant;

    try self.board_replace_history.append(self.board);

    const list_catchable = self.listCatchableOnce(marble_land_coord);
    errdefer list_catchable.deinit();

    if (list_catchable.items.len == 0) {
        self.current_player.changePlayer();
        self.game_state = if (self.whoIsWin()) |winner| .{ .game_end = winner } else blk: {
            break :blk .found_sequential_move;
        };
        self.sequential_move_list = if (self.sequential_move_list) |inner| blk: {
            inner.deinit();
            break :blk null;
        } else null;
        list_catchable.deinit();
    } else {
        self.game_state = if (self.whoIsWin()) |winner| .{ .game_end = winner } else blk: {
            break :blk .found_sequential_move;
        };
        self.sequential_move_list = if (self.sequential_move_list) |inner| blk: {
            inner.deinit();
            break :blk list_catchable;
        } else list_catchable;
    }
}

pub fn listAllCatchable(self: Self) GameError!ArrayList(CatchableMove) {
    var output = try ArrayList(CatchableMove).initCapacity(self.allocator, 81);
    errdefer output.deinit();

    var iter = CoordinateIter.new();
    while (iter.next()) |coord| {
        switch (self.board.getRaw(coord)) {
            .occupied => {
                const one_catchable = try self.listAllCatchableOnce(coord);
                defer one_catchable.deinit();
                try output.appendSlice(one_catchable.items);
            },
            else => {},
        }
    }

    return output;
}

fn listAllCatchableOnce(self: Self, coord: Coordinate) GameError!ArrayList(CatchableMove) {
    var output = try ArrayList(CatchableMove).initCapacity(self.allocator, 6);
    errdefer output.deinit();

    self.appendOccupiedCoordinate(&output, coord, Direction.up_right);
    self.appendOccupiedCoordinate(&output, coord, Direction.up);
    self.appendOccupiedCoordinate(&output, coord, Direction.left);
    self.appendOccupiedCoordinate(&output, coord, Direction.left_down);
    self.appendOccupiedCoordinate(&output, coord, Direction.down);
    self.appendOccupiedCoordinate(&output, coord, Direction.right);

    return output;
}

inline fn appendOccupiedCoordinate(
    self: Self,
    coord_list: *ArrayList(CatchableMove),
    coord: Coordinate,
    comptime direction: Direction,
) GameError!void {
    const catched_coord = coord.adjacent(direction) orelse return;
    const marble_land_coord = catched_coord.adjacent(direction) orelse return;

    if (self.board.get(catched_coord)) |catched_ring| {
        switch (catched_ring) {
            .occupied => {
                if (self.board.get(marble_land_coord)) |marble_land_ring| {
                    switch (marble_land_ring) {
                        .vacant => {
                            try coord_list.append(.{
                                .start_coord = coord,
                                .catched_coord = catched_coord,
                                .marble_land_coord = marble_land_coord,
                            });
                        },
                        else => {},
                    }
                }
            },
            else => {},
        }
    }
}

// ╭──────────────────────────────────────────────────────────╮
// │                    Putting Marble Api                    │
// ╰──────────────────────────────────────────────────────────╯

pub fn putMarble(
    self: *Self,
    put_coord: Coordinate,
    remove_coord: Coordinate,
    marble: Marble,
) GameError!void {
    if (self.board.getRaw(put_coord) == .vacant) {
        if (!self.total_marble.dec(marble) //
        and !self.players_score[@enumToInt(self.current_player)].dec(marble)) {
            return error.InvalidPuttingMarble;
        }

        self.board.getRawMut(put_coord).* = .{ .occupied = marble };
    } else {
        return error.InvalidPuttingMarble;
    }

    try self.removeRing(remove_coord);
    try self.removeIsolatedIsland();
    self.current_player.changePlayer();
    self.game_state = .check_is_catchable;
}

fn removeRing(self: *Self, coord: Coordinate) GameError!void {
    const list_removable = self.collectRemovableRings();
    defer list_removable.deinit();

    if (list_removable.items.len == 0) {
        if (!self.validToRemoveRing(coord)) return error.InvalidRingToRemove;

        self.board_replace_history.clearRetainingCapacity();
        self.repeat_count = 0;

        const ring = self.board.getRawMut(coord);
        if (ring.* != .vacant) return error.InvalidRingToRemove;
        ring.* = .empty;
        self.calculateCompoments();
    }
}

pub fn collectRemovableRings(self: Self) GameError!ArrayList(Coordinate) {
    var output = try ArrayList(Coordinate).initCapacity(self.allocator, 81);
    errdefer output.deinit();

    var iter = CoordinateIter.new();
    while (iter.next()) |coord| {
        if (self.validToRemoveRing(coord)) try output.append(coord);
    }

    return output;
}

inline fn validToRemoveRingHelper(
    self: Self,
    coord: Coordinate,
    comptime direction1: Direction,
    comptime direction2: Direction,
) bool {
    const coord1 = coord.adjacent(direction1);
    const coord2 = coord.adjacent(direction2);
    const ring1 = self.board.getOptional(coord1) orelse return false;
    const ring2 = self.board.getOptional(coord2) orelse return false;

    return ring1 == .empty and ring2 == .empty;
}

fn validToRemoveRing(self: Self, coord: Coordinate) bool {
    return self.validToRemoveRingHelper(coord, Direction.up_right, Direction.up) //
    or self.validToRemoveRingHelper(coord, Direction.left, Direction.up) //
    or self.validToRemoveRingHelper(coord, Direction.left, Direction.left_down) //
    or self.validToRemoveRingHelper(coord, Direction.down, Direction.left_down) //
    or self.validToRemoveRingHelper(coord, Direction.down, Direction.right) //
    or self.validToRemoveRingHelper(coord, Direction.up_right, Direction.right);
}

pub fn calculateCompoments(self: *Self) void {
    self.components.clear();

    var right_coord = undefined;
    var up_coord = undefined;
    var up_right_coord = undefined;
    var iter = CoordinateIter.new();

    while (iter.next()) |coord| {
        if (self.board.getRaw(coord).isEq(.empty)) {
            self.components.unionBoth(coord, main_empty_coord);
        }

        right_coord = coord.rawAdjacent(false, true, false, false);
        up_coord = coord.rawAdjacent(false, false, true, false);
        up_right_coord = coord.rawAdjacent(false, true, true, false);

        if (self.board.get(right_coord)) |ring| {
            switch (ring) {
                .empty => self.components.unionBoth(right_coord, main_empty_coord),
                else => if (ring.isEq(self.board.getRaw(coord))) blk: {
                    break :blk self.components.unionBoth(coord, right_coord);
                },
            }
        }

        if (self.board.get(up_coord)) |ring| {
            switch (ring) {
                .empty => self.components.unionBoth(up_coord, main_empty_coord),
                else => if (ring.isEq(self.board.getRaw(coord))) blk: {
                    break :blk self.components.unionBoth(coord, up_coord);
                },
            }
        }

        if (self.board.get(up_right_coord)) |ring| {
            switch (ring) {
                .empty => self.components.unionBoth(up_right_coord, main_empty_coord),
                else => if (ring.isEq(self.board.getRaw(coord))) blk: {
                    break :blk self.components.unionBoth(coord, up_right_coord);
                },
            }
        }
    }
}

fn removeIsolatedIsland(self: *Self) GameError!void {
    const main_components = try self.components.getComponentRepresentors();
    defer main_components.deinit();

    components: for (main_components.items) |main_coord| {
        if (self.board.getRaw(main_coord) == .empty) continue :components;

        var component_members = try ArrayList(Coordinate).initCapacity(self.allocator, 81);
        defer component_members.deinit();

        var iter = CoordinateIter.new();
        while (iter.next()) |coord| {
            if (self.components.find(main_coord) != self.components.find(coord)) continue;
            if (self.board.getRaw(coord) == .vacant) continue :components;

            try component_members.append(coord);
        }

        for (component_members.items) |coord| {
            switch (self.board.getRaw(coord)) {
                .occupied => |marble| self.players_score[@enumToInt(self.current_player)].inc(marble),
                else => unreachable,
            }

            self.board.getRawMut(coord).* = .empty;
            self.components.unionBoth(coord, main_empty_coord);
        }
    }
}

// ╭──────────────────────────────────────────────────────────╮
// │                     Check who is win                     │
// ╰──────────────────────────────────────────────────────────╯

fn whoIsWin(self: *Self) ?Player {
    if (self.players_score[@enumToInt(.alice)].isWin()) return .alice;
    if (self.players_score[@enumToInt(.bob)].isWin()) return .bob;

    for (self.board_replace_history.items) |prev_board| {
        if (meta.eql(self.board, prev_board)) {
            break;
        }
    } else {
        return null;
    }

    self.repeat_count += 1;
    if (self.repeat_count >= 3) return .tie;

    return null;
}
